### PR TITLE
Update LRScheduler to handle tensor LR

### PR DESCRIPTION
**Pull request by mlazos**
Enables LRScheduler to handle tensor LRs.

Note on test changes:
For the test modifications I just removed itertools.product and created two loops. This allows us to create a new set of optim_inputs on each iteration to prevent mutations on the tensor LR carrying over across iterations. Nothing else in those tests was modified. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #125653
* #125383
* __->__ #123753
* #123752
* #123751

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by mlazos** [comment]
@janeyx99 I think this one is ready as well.

**Comment by janeyx99** [comment]
Test failures are real

**Comment by mlazos** [comment]
> * test_forloop_goes_right_direction_contiguous_False_with_lrsched_True_SGD_cpu_float32

Should be fixed now, do the code changes look okay?

**Comment by mlazos** [comment]
> Test failures are real

Should be fixed now, do the code changes look okay?

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -1447,7 +1450,10 @@ def __init__(
         base_lrs = self._format_param("base_lr", optimizer, base_lr)
         if last_epoch == -1:
             for lr, group in zip(base_lrs, optimizer.param_groups):
-                group["lr"] = lr
+                if isinstance(group["lr"], Tensor):
+                    group["lr"].fill_(lr)
```
Comment (janeyx99) [inline torch/optim/lr_scheduler.py:1454]: are we guaranteed here that lr will never be a Tensor?

File: torch/optim/lr_scheduler.py [hunk 2]
```diff
@@ -182,7 +182,10 @@ def step(self, epoch=None):
 
         for i, data in enumerate(zip(self.optimizer.param_groups, values)):
             param_group, lr = data
-            param_group["lr"] = lr
+            if isinstance(param_group["lr"], Tensor):
+                param_group["lr"].fill_(lr)
```
Comment (janeyx99) [inline torch/optim/lr_scheduler.py:186]: same here

**Review by janeyx99 — COMMENTED** [review]
Overall this looks good to me; it looks like we are not supporting Tensor lr values throughout the LRScheduler instance itself--will that be fine for compile?

File: torch/optim/lr_scheduler.py (hunk 1 above)
Comment (mlazos) [reply torch/optim/lr_scheduler.py:1454]: I think we are, but I don't think there's harm in extracting the value out, let me do that.

**Review by janeyx99 — APPROVED** [review]

**Pull request closed without merging**